crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,328 @@
+    pub cool_down_interval_ms: u64,
+}
+
+impl Default for CoreFetchConfig {
+    fn default() -> Self {
+        Self {
+            parallel_request_count: 2,
+            cool_down_interval_ms: 120_000,
+        }
+    }
+}
+
+impl ModConfig for CoreFetchConfig {}
+
+// TODO: Temporary trait object of a transport module to facilitate unit tests.
+type DynTransport = Arc<dyn Transport>;
+
+#[derive(Debug)]
+struct CoreFetch(Inner);
+
+impl CoreFetch {
+    fn new(config: CoreFetchConfig, transport: DynTransport) -> Self {
+        let inner = Inner::new(config.cool_down_interval_ms);
+        inner.spawn_fetch_tasks(config.parallel_request_count, transport);
+        Self(inner)
+    }
+}
+
+impl Fetch for CoreFetch {
+    fn add_ops(
+        &mut self,
+        op_list: Vec<OpId>,
+        source: AgentId,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Add ops to set.
+            let ops = &mut self.0.state.lock().await.ops;
+            ops.extend(
+                op_list
+                    .clone()
+                    .into_iter()
+                    .map(|op_id| (op_id.clone(), source.clone())),
+            );
+
+            // Pass ops to fetch tasks.
+            op_list.into_iter().for_each(|op_id| {
+                if let Err(err) =
+                    self.0.fetch_queue_tx.try_send((op_id, source.clone()))
+                {
+                    tracing::warn!(
+                        "could not pass fetch request to fetch task: {err}"
+                    );
+                }
+            });
+
+            Ok(())
+        })
+    }
+}
+
+type FetchRequest = (OpId, AgentId);
+
+#[derive(Debug)]
+struct State {
+    ops: HashSet<FetchRequest>,
+    cool_down_list: CoolDownList,
+}
+
+#[derive(Debug)]
+struct Inner {
+    state: Arc<Mutex<State>>,
+    fetch_queue_tx: Sender<FetchRequest>,
+    fetch_queue_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+}
+
+impl Inner {
+    pub fn new(cool_down_interval_ms: u64) -> Self {
+        // Create a channel to send new ops to fetch to the tasks. This is in effect the fetch queue.
+        let (fetch_queue_tx, fetch_queue_rx) = channel::<FetchRequest>(1024);
+        let fetch_queue_rx = Arc::new(Mutex::new(fetch_queue_rx));
+
+        let ops = HashSet::new();
+        let cool_down_list = CoolDownList::new(cool_down_interval_ms);
+
+        Self {
+            state: Arc::new(Mutex::new(State {
+                ops,
+                cool_down_list,
+            })),
+            fetch_queue_tx,
+            fetch_queue_rx,
+        }
+    }
+
+    pub fn spawn_fetch_tasks(
+        &self,
+        parallel_request_count: u8,
+        transport: DynTransport,
+    ) {
+        for _ in 0..parallel_request_count {
+            tokio::task::spawn(Inner::fetch_task(
+                self.fetch_queue_tx.clone(),
+                self.fetch_queue_rx.clone(),
+                transport.clone(),
+                self.state.clone(),
+            ));
```

Comment:
The join handles of these task should be tracked in your top-level struct like `Vec<tokio::task::JoinHandle<()>>` so that they can be aborted on drop:

```rust
impl Drop for Inner /* or CoreFetch */ {
    fn drop(&mut self) {
        for t in self.task_list.iter() {
            t.abort();
````

Reply:
Alright! https://github.com/holochain/kitsune2/pull/35/commits/2a156d55b84470e058faf4366bbdd3a8b71a063e